src/pkg/boltz.ts

```diff
@@ -0,0 +1,480 @@
+}
+
+interface CreateSwapCommonRequest {
+  type: 'submarine' | 'reversesubmarine';
+  pairId: 'L-BTC/BTC';
+  orderSide: 'buy' | 'sell';
+}
+
+interface CreateSwapCommonResponse {
+  id: string;
+  timeoutBlockHeight: number;
+}
+
+interface SubmarineSwapRequest {
+  invoice: string;
+  refundPublicKey: string;
+}
+
+interface SubmarineSwapResponse {
+  acceptZeroConf: boolean;
+  address: string;
+  bip21: string;
+  blindingKey: string; // private
+  expectedAmount: number;
+  redeemScript: string;
+}
+
+interface ReverseSubmarineSwapRequest {
+  preimageHash: string;
+  invoiceAmount: number;
+  claimPublicKey: string;
+}
+
+interface ReverseSubmarineSwapResponse {
+  blindingKey: string; // private
+  invoice: string;
+  lockupAddress: string;
+  onchainAmount: number;
+  redeemScript: string;
+}
+
+export interface SubmarineSwap {
+  address: string;
+  expectedAmount: number;
+  id: string;
+  redeemScript: string;
+}
+
+export interface ReverseSwap {
+  blindingPrivateKey: string;
+  claimPublicKey: string;
+  id: string;
+  invoice: string;
+  lockupAddress: string;
+  preimage: Buffer;
+  redeemScript: string;
+  timeoutBlockHeight: number;
+}
+
+export interface MakeClaimTransactionParams {
+  utxo: Unspent;
+  claimKeyPair: ECPairInterface;
+  preimage: Buffer;
+  redeemScript: Buffer;
+  destinationScript: Buffer;
+  fee: number;
+  blindingPublicKey: Buffer;
+  timeoutBlockHeight: number;
+}
+
+const feeAmount = 500; // fee for regular liquid tx
+const swapFeeAmount = 500; // fee for Boltz
+// lightning swap invoice amount limit (in satoshis)
+export const DEFAULT_LIGHTNING_LIMITS = { maximal: 4294967, minimal: 50000 };
+export const DEPOSIT_LIGHTNING_LIMITS = {
+  maximal: DEFAULT_LIGHTNING_LIMITS.maximal - feeAmount - swapFeeAmount,
+  minimal: DEFAULT_LIGHTNING_LIMITS.minimal - feeAmount - swapFeeAmount,
+};
+
+export const boltzUrl: Record<NetworkString, string> = {
+  regtest: 'http://localhost:9090',
+  testnet: 'https://testnet.boltz.exchange/api',
+  liquid: 'https://api.boltz.exchange',
+};
+
+export class Boltz implements BoltzInterface {
+  private asset: string;
+  private url: string;
+  private zkp: ZKP;
+
+  constructor(url: string, asset: string, zkp: ZKP) {
+    this.asset = asset;
+    this.url = url;
+    this.zkp = zkp;
+  }
+
+  async getBoltzPair(pair: string): Promise<any> {
+    const data = await this.getApi(`${this.url}/getpairs`);
+    if (!data?.pairs?.[pair]) return;
+    return data.pairs[pair];
+  }
+
+  // return invoice expire date
+  getInvoiceExpireDate(invoice: string): number {
+    const toMilliseconds = (num: number) => num * 1000;
+    const { timeExpireDate, timestamp } = bolt11.decode(invoice);
+    if (timeExpireDate) return toMilliseconds(timeExpireDate);
+    const { expire_time } = bolt11.decode(invoice).tagsObject;
+    const expireTime = toMilliseconds(expire_time ?? 3600);
+    const from = timestamp ? toMilliseconds(timestamp) : Date.now();
```

Comment:
https://github.com/lightning/bolts/blob/master/11-payment-encoding.md

Timestamps are mandatory

> timestamp: seconds-since-1970 (35 bits, big-endian)

You could and probably should throw an error in case there is no timestamp

Reply:
Fixed by efce217e